File: 백준/Gold/1520. 내리막 길/내리막 길.java
import java.util.Scanner;

public class Main {

    private static final int[] dx = {1, -1, 0, 0};

    private static final int[] dy = {0, 0, 1, -1};

    private static int[][] graph;

    private static int[][] dp;

    public static void main(String[] args) {
        try (Scanner scanner = new Scanner(System.in)) {
            int m = scanner.nextInt();
            int n = scanner.nextInt();

            graph = new int[m][n];
            dp = new int[m][n];

            for (int i = 0; i < m; i++) {
                for (int j = 0; j < n; j++) {
                    graph[i][j] = scanner.nextInt();
                    dp[i][j] = -1;
                }
            }

            System.out.println(dfs(0, 0));
        }
    }

    private static int dfs(int x, int y) {
        if (x == graph[0].length - 1 && y == graph.length - 1) {
            return 1;
        }

        if (dp[y][x] != -1) {
            return dp[y][x];
        }

        dp[y][x] = 0;

        for (int i = 0; i < 4; i++) {
            int cx = x + dx[i];
            int cy = y + dy[i];

            if (cx >= 0 && cy >= 0 && cx < graph[0].length && cy < graph.length && graph[cy][cx] < graph[y][x]) {
                dp[y][x] += dfs(cx, cy);
            }
        }

        return dp[y][x];
    }
}

File: 백준/Gold/1520. 내리막 길/내리막 길.kt
val dx = arrayOf(1, -1, 0, 0)
val dy = arrayOf(0, 0, 1, -1)

fun main() {
    val (m, n) = readln().split(" ").map { it.toInt() }
    val graph = Array(m) { IntArray(n) }
    val dp = Array(m) { IntArray(n) { -1 } }

    for (i in 0 until m) {
        val input = readln().split(" ").map { it.toInt() }
        input.forEachIndexed { index, value ->
            graph[i][index] = value
        }
    }

    println(dfs(0, 0, graph, dp))
}

private fun dfs(
    x: Int,
    y: Int,
    graph: Array<IntArray>,
    dp: Array<IntArray>
): Int {
    if (x == graph.size - 1 && y == graph[0].size - 1) return 1
    if (dp[x][y] != -1) return dp[x][y]

    dp[x][y] = 0
    for (i in 0..3) {
        val nx = x + dx[i]
        val ny = y + dy[i]

        if (nx in graph.indices && ny in graph.first().indices && graph[x][y] > graph[nx][ny]) {
            dp[x][y] += dfs(nx, ny, graph, dp)
        }
    }

    return dp[x][y]
}